File: lectrick/tiles/horizontal_wire.py
from .one_way_base import OneWayTile


class HorizontalWire(OneWayTile):
    TILE_TYPE = "HORIZONTAL WIRE"
    SAMPLE_CHARS = ["-", "‐", "‒"]

    def __init__(self, x, y, tiles):
        super().__init__(x, y, tiles)

    def acceptance_filter(self, source_x, source_y):
        if source_y != self.y:
            return False
        return True

    def start_turn(self, *args, **kwargs):
        if self.current_input:
            amount, source_x, source_y = self.current_input
            if source_x < self.x:
                dx = -1
            elif source_x > self.x:
                dx = 1
            else:
                return
            try:
                tile_index = self.tiles.index((self.x - dx, self.y))
            except ValueError:
                return
            self.tiles[tile_index].accept_energy(amount, self.x, self.y)

File: web/src/tiles/horizontalWire.ts
import {OneWayBase} from './oneWayBase';

export class HorizontalWireTile extends OneWayBase {
    TILE_TYPE = 'HORIZONTAL WIRE';
    SAMPLE_CHARS = ["-", "‐", "‒"];

    constructor(ctx: ExecutionContext, x: number, y: number) {
        super(ctx, x, y);
    }

    acceptanceFilter(x: number, y: number): boolean {
        return y === this.y
    }

    startTurn() {
        if (!this.currentInput) {
            return;
        }
        const [amount, x, y] = this.currentInput;
        let dx;
        if (x < this.x) {
            dx = -1; // Left
        } else if (x > this.x) {
            dx = 1; // Right
        } else {
            return; // No horizontal movement
        }
        const targetTile = this.ctx.getTile(this.x + dx, this.y);
        if (targetTile) {
            targetTile.acceptEnergy(amount, this.x, this.y);
        }
    }
}